### DISCUSSION TITLE
Using ruby's irb in repl  or console

### DISCUSSION THREAD

**Discussion by phallguy** [mfussenegger/nvim-dap#1284] [Q&A]
Loving the debugger integration so far. JS/TS setup is working amazing and I've got ruby mostly working. However one of the features I miss the most from direct console debugging is Ruby's [IRB feature](https://github.com/ruby/irb). Is there any way to get a more interactive repl/console via the debugger protocol so that I can have the best of both worlds? 

The features I miss the most are some of the inspection features that let me interrogate types and values a little easier than the current repl or dap-ui allows.

Thanks for any help or insights!

**Comment by mfussenegger** [comment]
I'm not sure what exactly you're looking for as I'm not familiar with Ruby's IRB feature.

The debug adapter protocol itself has no mechanism specifically to run another debugger CLI in parallel in some collaborate mode, but some debug adapter support pass-through mode of their regular commands (e.g. gdb and codelldb have options for that)

> The features I miss the most are some of the inspection features that let me interrogate types and values a little easier than the current repl or dap-ui allows.

If you can describe these in more detail, maybe there are ways to add these to nvim-dap more directly.
I have in mind to tweak the REPL a bit anyway and learning more about common use-cases and functionality people want to use would be good.

**Reply by phallguy** [reply]
Awesome, this tool has already been super helpful for me in my process.

As for Ruby, the interactive repl offers a few inspection tools that I find really useful.

1. Objects can define an #inspect method that is intended for use when debugging to get a picture of the current state of the object. Its often formatted as a key/value list in the console.

2. irb has an 'ls' command that I use a lot. It shows me all of the methods on an object as well as instance variables. I use it a lot when trying to understand a 3rd party package and figure out why I'm using it wrong :)

This may be limited to Ruby's implementation of DAP, but I find it more cumbersome to interrogate the state of an object with either dap-ui or the repl. I find it has either too little or too much info. For example large strings are truncated, as well as arrays. Or I'll get the entire class tree...when I really want a key/value list of data - similar to how chrome dev tools lets you drill into objects.

Of course this could all just be my limited understanding of the tools and I'm just missing something silly. I did find that prefixing the commands with a comma let me access native IRB. So woot! But for some reason its really slow and I sometimes have to try the command multiple times before it'll run.

Thanks again for all the work you've already done and for taking the time to respond!
